### PR TITLE
Add quality metrics integration to diagnose_and_remediate_data.py

### PR DESCRIPTION
**Pull request by sk413025**
This PR adds integration with the quality_metrics.py module in the data diagnostics utilities.

## Changes Made

1. Add imports for quality metrics functions from utils.data_diagnostics
2. Add command-line arguments --quality-metrics and --num-samples to control the quality metrics analysis
3. Add a comprehensive function for calculating and storing quality metrics for data samples
4. Update the main function to conditionally use the quality metrics analysis when requested

## Details

The implementation now allows users to analyze samples using all six quality metrics functions:
- calculate_sample_difficulty
- calculate_sample_influence
- calculate_feature_space_density
- calculate_prediction_stability
- calculate_loss_landscape
- calculate_comprehensive_quality_score

Example usage:
```
python scripts/diagnose_and_remediate_data.py --frequency 1000hz --material plastic --quality-metrics --num-samples 5
```

The quality metrics results are saved in JSON format in the logs directory for further analysis.

### CODE REVIEW

**Comment by sk413025** [comment]
## Technical Implementation Details

This PR completes an important part of the data diagnostics module by integrating all the quality metrics functions into the diagnostics workflow. Here are the technical details of the implementation:

### Data Shape Handling
One of the key technical challenges was ensuring proper shape handling across different data formats. The implementation addresses this by:

1. Adding automatic shape detection and transformation for input tensors:
   ```python
   # 確保有正確的維度 (假設模型期望 N,C,H,W 格式)
   if len(sample.shape) == 3:  # (C,H,W)
       sample = sample.unsqueeze(0)  # 添加批次維度: (1,C,H,W)
   elif len(sample.shape) == 2:  # (H,W)
       sample = sample.unsqueeze(0).unsqueeze(0)  # 添加通道和批次維度: (1,1,H,W)
   ```

2. Processing validation set samples with the same shape transformation logic

### Advanced Quality Metrics Implementation
The implementation utilizes all six quality metrics functions:

1. **Sample Difficulty** - Measures how difficult samples are for the model to predict correctly
2. **Sample Influence** - Evaluates how much a sample affects model performance when used in training
3. **Feature Space Density** - Analyzes the sample's position in feature space relative to other samples
4. **Prediction Stability** - Tests how stable model predictions are under small input perturbations
5. **Loss Landscape** - Analyzes the loss surface around the sample to detect problematic curvatures
6. **Comprehensive Quality Score** - Combines multiple metrics for an overall assessment

### Error Handling and Robustness
The implementation incorporates robust error handling for each metric calculation:

```python
try:
    difficulty = calculate_sample_difficulty(model, sample, target)
    print(f"  - 樣本難度: {difficulty:.4f}")
except Exception as e:
    print(f"  - 計算樣本難度時出錯: {e}")
    difficulty = None
```

This ensures that failures in one metric don't stop the entire analysis process.

### Data Augmentation for Stability Analysis
A custom implementation of data augmentation functions was added for prediction stability testing:

```python
def add_noise(x, noise_level=0.05):
    return x + torch.randn_like(x) * noise_level
    
augmentations = [
    lambda x: add_noise(x, 0.01),
    lambda x: add_noise(x, 0.05),
    lambda x: add_noise(x, 0.1)
]
```

### JSON Serialization
The implementation carefully handles serialization of complex nested structures and torch tensors to ensure proper JSON output:

```python
json_results = {}
for idx, metrics in quality_results.items():
    json_results[str(idx)] = {}
    for metric_name, value in metrics.items():
        if metric_name == 'loss_landscape' and value is not None:
            # Only save key metrics from loss landscape
            json_results[str(idx)][metric_name] = {
                'curvature': float(value['curvature']),
                'original_loss': float(value['original_loss']),
                ...
            }
```

This PR fills a significant gap in our diagnostic toolset, enabling detailed quality analysis of individual samples in our datasets.

This addresses part of issue #28 related to implementing data quality metrics.
